### DISCUSSION TITLE
Custom Blocks for Angular project

### DISCUSSION THREAD

**Discussion by TimGeerts** [storybookjs/storybook#26380] [Help] [answered]
### Summary

I'm trying to reuse some "returning" paragraphs in my documentation, and I bumped into the `useOf` [documentation ](https://storybook.js.org/docs/api/doc-block-useof).  
Now, sadly, this documentation is written in jsx (nothing changes if I swap to "Angular" on the top.  
But I decided to just use the example for the `StoryName` from that page, and add it to a regular `.js` file, wrapping the returning `html` code in quotes.

It "kinda" works, but it also very much doesn't.
If I for instance just return like this:  

```
...
return '<h1>this is a test</h1>';
...
```

Then I'll just get that on my documentation as a string, including the ` h-tags`, so it'll look something like this:

![image](https://github.com/storybookjs/storybook/assets/2218719/50a6c026-6b10-430b-a50f-1398e375e5b4)

Is there a way to have the actual html being rendered on the page in stead?  
I assume the react way (no ticks) works because some plugin will switch on and will render the returning value from a jsx file.

### Additional information

main.js file
Storybook v7.5.x

```
const config = {
  stories: ['../**/src/**/*.mdx', '../**/src/**/*.stories.@(js|jsx|ts|tsx|mdx)'],
  addons: [
    '@storybook/addon-essentials',
    '@storybook/addon-docs',
    '@storybook/addon-a11y',
    '@storybook/addon-links',
    '@storybook/addon-actions',
    'storybook-addon-designs',
    'storybook-addon-angular-router',
    {
      name: '@storybook/addon-styling',
      options: {
        sass: {
          implementation: require('sass'),
        },
      },
    },
  ],
  framework: {
    name: '@storybook/angular',
    options: {},
  },
  staticDirs: ['./assets'],
};

export default config;
```

**Comment by Marklb** [comment, marked as answer]
I haven't used JSX in an Angular Storybook in a while, since my project has some other things to fix before I get back to that, but I think you just need to add `{ "compilerOptions": { ..., "jsx": "react" }` to your tsconfig for JSX syntax to work. To test that it still works, I setup a Stackblitz project and copied the `StoryName` code. It was a little bit of trouble getting it to work, so I would need to spend some time undoing the changes to see what changes are necessary, but you could take a look at it as an example. If I have some time, I will figure out what changes are necessary and update this answer. Also, I had trouble using `StoryName.jsx`, but `StoryName.tsx` seemed fine. Stackblitz: https://stackblitz.com/edit/github-i2mhyh

If you don't want to use jsx then you could convert the jsx to js and try that.

Here is the `StoryName` in js:

```js
import { useOf } from '@storybook/blocks';

/**
 * A block that displays the story name or title from the of prop
 * - if a story reference is passed, it renders the story name
 * - if a meta reference is passed, it renders the stories' title
 * - if nothing is passed, it defaults to the primary story
 */
export const StoryName = ({ of }) => {
  const resolvedOf = useOf(of || 'story', ['story', 'meta']);
  switch (resolvedOf.type) {
    case 'story': {
      return React.createElement('h1', null, resolvedOf.story.name);
    }
    case 'meta': {
      return React.createElement('h1', null, resolvedOf.preparedMeta.title);
    }
  }
  return null;
};
```

---
Unrelated to your question, but is there a reason you are requiring 'sass' with `@storybook/addon-styling`? The Angular builder already has a sass loader, so you are most likely overwriting Angular's loader and may run into problems with how various things are handled in your stylesheets. Most styling issues that I used to help with were caused by how the old Storybook docs were a bit misleading and made it sound like you needed to add a loader for preprocessors.

**Reply by TimGeerts** [reply]
I tried your approach, adding the compiler options and using a tsx file and I got it to work!!!
Thanks a lot.

For future reference, these are the important things I did to make it work:
- add `{ "compilerOptions": { ..., "jsx": "react" }` to the `tsconfig` file of my .storybook folder
- added the example `StoryName.tsx` file to `.storybook/blocks/` ([this example](https://storybook.js.org/docs/api/doc-block-useof))
  - to make that example a valid `tsx` file from the `jsx` displayed on the site I added `import * as React from 'react';` at the top and gave the `of` parameter an `any` type (`({ of }: { of: any }) => {`)
- added a new "includes" rule to the ´tsconfig´ file of my .storybook folder (`"../.storybook/blocks/*.tsx"`)

And then I could import and use that custom block in my `.mdx` file.

Thanks!

**Comment by TimGeerts** [comment]
Thanks, I'll try your approach with the examples.
About the sass, our storybook requires loading global styles form another lib we have (monorepo with NX), and it's been there since we were on v6 of Storybook.
I can see now from docs that it might not be needed any longer since v7, but if I try to remove it all, I keep running into webpack mistakes, the latest one being "Unexpected character '@'", so I don't think I can remove it that easily...

**edit**: I forgot to move my imports from the sass files to the project.json file of the storybook config, I removed the `@storybook/addon-styling` configuration from my main.js file and removed the sass imports from the preview.js file and it now works.
Thanks for the tip!
